controller/system_rollout_controller.go

```diff
@@ -0,0 +1,1801 @@
+			types.KubernetesKindClusterRoleList:           c.restoreClusterRoles,
+			types.KubernetesKindClusterRoleBindingList:    c.restoreClusterRoleBindings,
+			types.KubernetesKindPodSecurityPolicyList:     c.restorePodSecurityPolicies,
+			types.KubernetesKindRoleList:                  c.restoreRoles,
+			types.KubernetesKindRoleBindingList:           c.restoreRoleBindings,
+			types.KubernetesKindConfigMapList:             c.restoreConfigMaps,
+			types.KubernetesKindDeploymentList:            c.restoreDeployments,
+			types.KubernetesKindDaemonsetList:             c.restoreDaemonSets,
+			types.LonghornKindEngineImageList:             c.restoreEngineImages,
+			types.LonghornKindVolumeList:                  c.restoreVolumes,
+			types.KubernetesKindPersistentVolumeList:      c.restorePersistentVolumes,
+			types.KubernetesKindPersistentVolumeClaimList: c.restorePersistentVolumeClaims,
+			types.LonghornKindRecurringJobList:            c.restoreRecurringJobs,
+		}
+		wg.Add(len(restoreFns))
+		for k, v := range restoreFns {
+			kind, fn := k, v
+			go func() {
+				c.restore(kind, fn, log)
+				wg.Done()
+			}()
+		}
+		wg.Wait()
+
+		if len(c.cacheErrors) == 0 {
+			c.cacheStateUpdate(
+				systemRolloutRecordTypeNormal, longhorn.SystemRestoreStateCompleted,
+				EventReasonRestored, SystemRolloutMsgCompleted,
+			)
+		}
+	}
+
+	return nil
+}
+
+func (c *SystemRolloutController) initializeSystemRollout(systemRestore *longhorn.SystemRestore, log logrus.FieldLogger) (*longhorn.SystemRestore, error) {
+	systemRestore.Status.OwnerID = c.controllerID
+	systemRestore.Status.State = longhorn.SystemRestoreStateInitializing
+
+	systemBackupURL, err := c.GetSystemBackupURL()
+	if err != nil {
+		return nil, err
+	}
+	systemRestore.Status.SourceURL = systemBackupURL
+
+	systemRestore, err = c.ds.UpdateSystemRestore(systemRestore)
+	if err != nil {
+		return nil, errors.Wrap(err, SystemRolloutErrUpdateSystemRestore)
+	}
+
+	return systemRestore, nil
+}
+
+func (c *SystemRolloutController) syncController() error {
+	if c.systemRestore == nil {
+		systemRestore, err := c.ds.GetSystemRestore(c.systemRestoreName)
+		if err != nil {
+			return err
+		}
+
+		c.downloadPath = filepath.Join(types.SystemRolloutDirTemp, systemRestore.Name+types.SystemBackupExtension)
+
+		systemBackup, err := c.ds.GetSystemBackupRO(systemRestore.Spec.SystemBackup)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get SystemBackup %v", systemRestore.Spec.SystemBackup)
+		}
+
+		c.systemRestoreVersion = systemBackup.Status.Version
+		c.systemRestore = systemRestore
+	}
+
+	backupTarget, err := c.ds.GetDefaultBackupTargetRO()
+	if err != nil {
+		return err
+	}
+
+	backupTargetClient, err := getBackupTargetClient(c.ds, backupTarget, "")
+	if err != nil {
+		return errors.Wrapf(err, "failed to init backup target clients")
+	}
+
+	systemBackupCfg, err := backupTargetClient.SystemBackupConfigGet(c.systemRestore.Spec.SystemBackup, c.systemRestoreVersion)
```

Comment:
If there is a mismatching between the recorded version and the current Longhorn version, it's better to have a warning log.

Reply:
Addressed.